lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,377 @@
+                             RTATestBundle, CannotCreateError)
+from lisa.target import Target
+from lisa.utils import ArtifactPath, groupby, memoized
+from lisa.wlgen.rta import RTA, RTATask, Periodic, Ramp
+from lisa.trace import FtraceConf, FtraceCollector, requires_events
+from lisa.analysis.rta import RTAEventsAnalysis
+from lisa.analysis.tasks import TaskState, TasksAnalysis
+from lisa.analysis.load_tracking import LoadTrackingAnalysis
+from lisa.datautils import series_integrate
+
+from lisa.tests.scheduler.load_tracking import LoadTrackingHelpers
+
+class UtilTrackingBase(RTATestBundle, LoadTrackingHelpers):
+    """
+    Base class for shared functionality of utilization tracking tests
+    """
+
+    cpufreq_conf = {"governor": "performance"}
+    """
+    The cpufreq configuration used while the synthetic workload is being run.
+    Items are arguments to :meth:`devlib.cpufreq.use_governor`.
+    """
+
+    @classmethod
+    def _from_target(cls, target:Target, *,
+                     res_dir:ArtifactPath=None,
+                     ftrace_coll:FtraceCollector=None) -> 'UtilTrackingBase':
+        plat_info = target.plat_info
+        rtapp_profile = cls.get_rtapp_profile(plat_info)
+
+        # After a bit of experimenting, it turns out that on some platforms
+        # misprediction of the idle time (which leads to a shallow idle state,
+        # a wakeup and another idle nap) can mess up the duty cycle of the
+        # rt-app task we're running. In our case, a 50% duty cycle, 16ms period
+        # task would always be active for 8ms, but it would sometimes sleep for
+        # only 5 or 6 ms.
+        # This is fine to do this here, as we only care about the proper
+        # behaviour of the signal on running/not-running tasks.
+        with target.disable_idle_states():
+            with target.cpufreq.use_governor(**cls.cpufreq_conf):
+                cls._run_rtapp(target, res_dir, rtapp_profile, ftrace_coll)
+
+        return cls(res_dir, plat_info)
+
+    @LoadTrackingAnalysis.df_tasks_signals.used_events
+    @requires_events('sched_switch', 'sched_util_est_cfs', 'sched_util_est_task')
+    def get_task_sched_signals(self, trace, cpu, task_name):
+        """
+        Get a :class:`pandas.DataFrame` with the sched signals for the workload task
+
+        This examines scheduler utilization tracking trace events. You will need a
+        target kernel that includes the required events.
+
+        :returns: :class:`pandas.DataFrame` with a column for each signal for
+          the workload task
+        """
+        df = trace.analysis.load_tracking.df_tasks_signals()
+        df = df[df['comm'] == task_name]
+        window = self.get_task_window(trace, task_name, cpu)
+        df = select_window(df, window)
+
+        # Normalize the signal with the detected task execution start
+        df.index -= window[0]
+
+        return df
+
+    @classmethod
+    def check_from_target(cls, target):
```

Comment:
Redefining `check_from_target()` is unnecessary, `from_target()` will already fail with an explicit error so that's just information duplication. `check_from_target` can be useful to check some emerging aspects of the system that will not make the test fail in an obvious way, but that's not the case here.

Reply:
That's gone